File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/merge_intervals.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

// O(nlogn) / O(1)
vector<Interval> merge_intervals(vector<Interval> &a) {
  sort(a.begin(), a.end(),
       [](auto &e1, auto &e2) { return e1.start <= e2.start; });
  vector<Interval> res;
  int start = a[0].start;
  int end = a[0].end;

  for (int i = 1; i < a.size(); i++) {
    if (a[i].start <= end) {
      end = max(end, a[i].end);
    } else {
      res.push_back(Interval(start, end));
      start = a[i].start;
      end = a[i].end;
    }
  }

  res.push_back(Interval(start, end));

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, start, end;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    for (int i = 0; i < n; i++) {
      cin >> start >> end;
      a.push_back(Interval(start, end));
    }

    auto res = merge_intervals(a);
    for (auto r : res) {
      cout << r.start << " " << r.end << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/MergeIntervals.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MergeIntervals {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
          int start = in.nextInt();
          int end = in.nextInt();
          a.add(new Interval(start, end));
        }

        List<Interval> res = mergeIntervals(a);
        for (var r : res) {
          System.out.printf("%d %d\n", r.start, r.end);
        }
      }
    }
  }

  // O(nlogn) / O(1)
  public static List<Interval> mergeIntervals(List<Interval> a) {
    a.sort((x, y) -> Integer.compare(x.start, y.start));

    List<Interval> res = new ArrayList<>();
    int start = a.get(0).start;
    int end = a.get(0).end;

    for (int i = 1; i < a.size(); i++) {
      if (a.get(i).start <= end) {
        end = Math.max(end, a.get(i).end);
      } else {
        res.add(new Interval(start, end));
        start = a.get(i).start;
        end = a.get(i).end;
      }
    }

    res.add(new Interval(start, end));

    return res;
  }
}